Net: Massive speedup. Net locks overhaul

theuni opened the pull request:
Depends on (includes) #9289. This is what I ultimately set out to fix with the net refactor.

In my (short) tests, it cuts network latency to ~50%. In some cases, more like 30%.

Test method: 1 fresh node (macbook) connected to a single other node (desktop). Testnet. Running until synced to block 10000. Results:
```
new = patched with this PR.
old = running #9289.
client / server
old / old: 100000 in 8:05
new / old: 100000 in 3:24
new / new: 100000 in 2:16
```
The results are very reproducible, always within a few seconds. Not only is it a nice improvement for the fresh node, but it compounds when its peer is running the updated code as well.

I had hoped to have the abstraction complete in time for 0.14, but it looks like that's unlikely at this point. For reference, it would look something more like https://github.com/theuni/bitcoin/commit/1a6b10aea129ac363727c2d68fae809c2861e4da.

Anyway, this is an attempt to fix the actual issue in time for 0.14, and putting off the last bit of refactor until after that. This addresses the issue observed in #9415, but also cleans up the nasty locking issues.

I labeled this WIP because there are probably still several racy bits. Quite a bit of test-writing remains.

See the individual commits for the details. tl;dr: We currently either process a peer's message or read from their socket, but never both simultaneously. The changes here remove that restriction.

rebroad commented on src/net_processing.cpp:
```diff
@@ -2438,48 +2438,53 @@ bool ProcessMessages(CNode* pfrom, CConnman& connman)
 
-    std::deque<CNetMessage>::iterator it = pfrom->vRecvMsg.begin();
-    while (!pfrom->fDisconnect && it != pfrom->vRecvMsg.end()) {
         // Don't bother if send buffer is too full to respond anyway
```
I'm not sure if github is messing with the formatting, but this line looks unnecessarily indented.

theuni replied:
Yes, the indentation was just kept to avoid creating a huge whitespace diff. This can be cleaned up in a move-only commit as a next step.

theuni commented:
@sipa I've spent all day on changing this around some and breaking this up into logical commits in order to satisfy myself (and i hope others) that it's safe. I'll push up a fresh version in a few min, and drop the WIP tag.

theuni commented:
Updated. Don't let the number of commits scare you off, the diffstat isn't too bad.

I attempted to break this up into small/simple commits, in order to slowly remove the need for cs_vRecvMsg. The last commit actually removes it.

I'm satisfied that this shouldn't be introducing any new races, as only a handful of vars were actually touched on the SocketHandler thread.

theuni committed:
net: fix typo causing the wrong receive buffer size

Surprisingly this hasn't been causing me any issues while testing, probably
because it requires lots of large blocks to be flying around.

Send/Recv corks need tests!

theuni committed:
net: make vRecvMsg a list so that we can use splice()

theuni committed:
net: make GetReceiveFloodSize public

This will be needed so that the message processor can cork incoming messages

theuni committed:
net: only disconnect if fDisconnect has been set

These conditions are problematic to check without locking, and we shouldn't be
relying on the refcount to disconnect.

theuni committed:
net: wait until the node is destroyed to delete its recv buffer

when vRecvMsg becomes a private buffer, it won't make sense to allow other
threads to mess with it anymore.

theuni committed:
net: set message deserialization version when it's actually time to deserialize

We'll soon no longer have access to vRecvMsg, and this is more intuitive anyway.

theuni committed:
net: handle message accounting in ReceiveMsgBytes

This allows locking to be pushed down to only where it's needed

Also reuse the current time rather than checking multiple times.

theuni committed:
net: record bytes written before notifying the message processor

theuni committed:
net: Add a simple function for waking the message handler

This may be used publicly in the future

theuni committed:
net: remove useless comments

theuni committed:
net: rework the way that the messagehandler sleeps

In order to sleep accurately, the message handler needs to know if _any_ node
has more processing that it should do before the entire thread sleeps.

Rather than returning a value that represents whether ProcessMessages
encountered a message that should trigger a disconnnect, interpret the return
value as whether or not that node has more work to do.

Also, use a global fProcessWake value that can be set by other threads,
which takes precedence (for one cycle) over the messagehandler's decision.

Note that the previous behavior was to only process one message per loop
(except in the case of a bad checksum or invalid header). That was changed in
PR #3180.

The only change here in that regard is that the current node now falls to the
back of the processing queue for the bad checksum/invalid header cases.

theuni committed:
net: add a new message queue for the message processor

This separates the storage of messages from the net and queued messages for
processing, allowing the locks to be split.

theuni committed:
net: add a flag to indicate when a node's process queue is full

Messages are dumped very quickly from the socket handler to the processor, so
it's the depth of the processing queue that's interesting.

The socket handler checks the process queue's size during the brief message
hand-off and pauses if necessary, and the processor possibly unpauses each time
a message is popped off of its queue.

theuni committed:
net: add a flag to indicate when a node's send buffer is full

Similar to the recv flag, but this one indicates whether or not the net's send
buffer is full.

The socket handler checks the send queue when a new message is added and pauses
if necessary, and possibly unpauses after each message is drained from its buffer.

theuni committed:
net: remove cs_vRecvMsg

vRecvMsg is now only touched by the socket handler thread.

The accounting vars (nRecvBytes/nLastRecv/mapRecvBytesPerMsgCmd) are also
only used by the socket handler thread, with the exception of queries from
rpc/gui. These accesses are not threadsafe, but they never were. This needs to
be addressed separately.

Also, update comment describing data flow

theuni commented on src/net.cpp:
```diff
@@ -1243,9 +1243,10 @@ void CConnman::ThreadSocketHandler()
                             if(notify)
                                 condMsgProc.notify_one();
-                            RecordBytesRecv(nBytes);
+                            }
```
Whoops, rebase gone bad here. It ends up right in the end. Will fix.

theuni commented on src/net_processing.cpp:
```diff
@@ -2439,44 +2439,44 @@ bool ProcessMessages(CNode* pfrom, CConnman& connman, std::atomic<bool>& interru
 
     auto it = pfrom->vRecvMsg.begin();
-    while (!pfrom->fDisconnect && it != pfrom->vRecvMsg.end()) {
+    if (!pfrom->fDisconnect && it != pfrom->vRecvMsg.end()) {
```
Need to clarify in the commit message: It was almost always the case that only one message was processed in this while loop. See the break at line 2358.

This commit changes the behavior so that it's _always_ one message processed per-loop. The only messages that show different behavior are the ones that used to "continue" the loop. I think it's perfectly reasonable to skip to the next node for processing in that case.

TheBlueMatt replied:
Point is this is not a change. See https://github.com/bitcoin/bitcoin/pull/9441#issuecomment-270185545

dcousens commented:
@theuni is this worth testing now?

theuni commented:
@dcousens Very much so!

TheBlueMatt commented on src/net.cpp:
```diff
@@ -682,18 +674,27 @@ bool CNode::ReceiveMsgBytes(const char *pch, unsigned int nBytes, bool& complete
+    int64_t nTime = GetTimeMicros();
+    LOCK(cs_vRecvMsg);
+    nRecvBytes += nBytes;
+    nLastRecv = nTime * 1000;
```
I think you meant /, not * here?

theuni replied:
Heh, sure did

TheBlueMatt commented on src/net.cpp:
```diff
@@ -682,18 +674,27 @@ bool CNode::ReceiveMsgBytes(const char *pch, unsigned int nBytes, bool& complete
+
+    int64_t nTime = GetTimeMicros();
+    LOCK(cs_vRecvMsg);
+    nRecvBytes += nBytes;
```
It might be nice in the RPC for mapRecvBytesPerMsgCmd to add up to nRecvBytes (ie to increment this in the loop per-msg instead of after receiving (which would also mean no api change)).

theuni replied:
Sure, that just means more locking. Though this is completely uncontested now, so that's not really a problem.

TheBlueMatt commented:
See IRC discussion:

```
<BlueMatt> cfields: hmmm, regarding #9441, do we really want to run the entire ProcessMessages loop (calling SendMessages potentially umpteen times) just to process multiple messages from the same node?
<BlueMatt> (ie remove the loop inside ProcessMessages and move it to ThreadProcessMessages)
<sipa> BlueMatt: i was wondering about that too
<sipa> BlueMatt: but there is hardly any (contentious) locking going on anymore in between
<BlueMatt> yea, but SendMessages.....
<sipa> Ah.
<sipa> i hadn't considered that
<sipa> that defeats the purpose
<sipa> hmm, i was about to say that it negatively impacts batching of invs and addrs
<sipa> but since we have explicit random delays for those, i don't think that's really an issue anymore
<BlueMatt> yea, I dont think it breaks anything
<BlueMatt> just repeatedly calls SendMessages to do nothing
<sipa> oh, it won't break anything
```

theuni commented:
@TheBlueMatt If I understand you correctly, as mentioned in the summary, that's the part that was cut out here for simplicity. The next set of changes would combine the loops and move them into net_processing. See here for how it looked in a previous iteration: https://github.com/theuni/bitcoin/commit/1a6b10aea129ac363727c2d68fae809c2861e4da#diff-eff7adeaec73a769788bb78858815c91R271

I'd be happy to add that here, but I'm afraid it adds a significant review burden to this PR.

theuni commented:
Also from IRC:
```
<BlueMatt> I assume you didnt touch cs_vSend due to https://github.com/bitcoin/bitcoin/pull/9419/commits/c214d120a363a05ba9afdccff6b4bda6e29ae7c4, cfields?
<cfields> BlueMatt: yes, cs_vSend wasn't touched because of your PR. I've just been operating under the assumption that the lock around SendMessages will be removed by one PR or another.
```

theuni commented:
Digging further into the IRC conversation above, I'd like to clarify that the interaction here between ProcessMessages and SendMessages is _not_ changed substantially in this PR.

@TheBlueMatt had missed a subtle part of the current behavior, and maybe @sipa too, so I'd just like to explicitly point out the current break here: https://github.com/bitcoin/bitcoin/blob/master/src/net_processing.cpp#L2543, added in https://github.com/bitcoin/bitcoin/pull/3180

For the sake of processing in a round-robin fashion, we currently end up running through SendMessages multiple times while draining a node's received message queue. That behavior is _far_ from ideal, and needs to be addressed, but is not changed here*. I have plans to work on this in next steps, but not for 0.14.

*It's only changed in the case of an invalid header, or bad message checksum.

gmaxwell commented:
> <BlueMatt> cfields: hmmm, regarding #9441, do we really want to run the entire ProcessMessages loop (calling SendMessages potentially umpteen times) just to process multiple messages from the same node?
<BlueMatt> (ie remove the loop inside ProcessMessages and move it to ThreadProcessMessages)

Clearly people weren't reading PRs on this or you would have noticed that it only processes one at a time right now.  I actually fixed it to process as many as it could until it encountered lock contention, then specifically pinged people for input on the effect. After no response, I unfixed it because of concern about the latency impact.   Handling multiple messages at a time is good, if they're fast ones...

sipa commented:
I certainly noticed it only processed one message at a time. I missed that
it calls SendMessages between each.

theuni committed:
net: remove redundant max sendbuffer size check

This is left-over from before there was proper accounting. Hitting 2x the
sendbuffer size should not be possible.

sipa commented:
Needs rebase.

TheBlueMatt commented on src/net.cpp:
```diff
@@ -1051,8 +1051,7 @@ void CConnman::ThreadSocketHandler()
             BOOST_FOREACH(CNode* pnode, vNodesCopy)
             {
-                if (pnode->fDisconnect ||
-                    (pnode->GetRefCount() <= 0 && pnode->vRecvMsg.empty() && pnode->nSendSize == 0))
```
Is it just me, or were these always impossible to hit? Both ConnectNode and AcceptConnection do a AddRef() which is only countered by the Release a few lines further down here.

TheBlueMatt replied:
(obviously not an issue, just checking my logic).

theuni commented:
I believe you're correct, that's my reading of it as well. Regardless, I
wasn't able to track down a path where fDisconnect wouldn't be set as
necessary.

TheBlueMatt commented on src/net.cpp:
```diff
@@ -649,16 +644,19 @@ void CNode::copyStats(CNodeStats &stats)
+    LOCK(cs_vRecvMsg);
     complete = false;
+    int64_t nTime = GetTimeMicros();
+    nLastRecv = nTime / 1000;
```
I think you actually meant / 1000 / 1000 :p

theuni replied:
I completely missed that this was a full-second value. Thanks for catching this (twice!). Will fix.

TheBlueMatt commented on src/net_processing.cpp:
```diff
@@ -2453,44 +2453,44 @@ bool ProcessMessages(CNode* pfrom, CConnman& connman, std::atomic<bool>& interru
         {
             LogPrintf("PROCESSMESSAGE: ERRORS IN HEADER %s peer=%d\n", SanitizeString(hdr.GetCommand()), pfrom->id);
-            continue;
+            return fMoreWork;
```
This is always false here.

theuni replied:
This is unclear due to rebasing. In the end-result, this will return true if there are more messages.

In this commit, there's an interim chunk that's missing. line 2479 should have:
```c++
fMoreWork = !pfrom->vRecvMsg.empty()
```
I'll fix it in the next rebase so that it's more clear.

TheBlueMatt replied:
Yea, that line is just too far down in the function, it needs to be moved up.

pstratem replied:
This should set fDisconnect and return false unconditionally actually.

theuni replied:
Yes, that would be a change of behavior, and I'm not entirely sure that we want it changed.
Presumably this was done to avoid partitioning the network in the event that the header structure (or magic bytes) ever needs to change.

theuni commented on src/net.cpp:
```diff
@@ -1239,8 +1239,18 @@ void CConnman::ThreadSocketHandler()
-                            if (notify)
+                            if (notify) {
+                                auto it(pnode->vRecvMsg.begin());
+                                for (; it != pnode->vRecvMsg.end(); ++it) {
```
Sure, will change it to a while.

Iterating backwards would be preferable, but we end up hitting each msg anyway later (to add up message sizes) in a later commit, so i didn't bother (see https://github.com/bitcoin/bitcoin/pull/9441/commits/a109f7bd1801392bf99994a7165d495200156c4a#diff-9a82240fe7dfe86564178691cc57f2f1R1246). The sizes are added ahead of time in order to avoid iterating the list under cs_vProcessMsg.

theuni commented on src/net.cpp:
```diff
@@ -1882,10 +1878,9 @@ void CConnman::ThreadMessageHandler()
-            condMsgProc.wait_until(lock, std::chrono::steady_clock::now() + std::chrono::milliseconds(100));
-        }
+        std::unique_lock<std::mutex> lock(mutexMsgProc);
+        condMsgProc.wait_until(lock, std::chrono::steady_clock::now() + std::chrono::milliseconds(100), [this, &fMoreWork] { return fMoreWork || fMsgProcWake; });
```
Yep. Will fix.

TheBlueMatt commented on src/net.h:
```diff
@@ -650,6 +651,7 @@ class CNode
     const NodeId id;
 
     const uint64_t nKeyedNetGroup;
+    std::atomic_bool fPauseRecv;
```
This is never set to true until a later commit. I'm generally ok with this, but it is a bit strange.

theuni replied:
Grr, another artifact of breaking up an end-result into smaller commits. I'll add this where it belongs in the next rebase.

TheBlueMatt commented on src/net.cpp:
```diff
@@ -1239,8 +1239,18 @@ void CConnman::ThreadSocketHandler()
                             RecordBytesRecv(nBytes);
-                            if (notify)
+                            if (notify) {
+                                auto it(pnode->vRecvMsg.begin());
```
You're not holding cs_vRecvMsg here, which hyou need to access vRecvMsg.

theuni replied:
vRecvMsg no longer needs cs_vRecvMsg starting here. It's only ever used on this thread. In reality, cs_vRecvMsg isn't needed at all, so it would probably be more clear to just remove it. copyStats is the only thing that accesses some of these vars, but it already operates without any locks, so the situation isn't changed from before.

TheBlueMatt replied:
Hum, ok, can you just add a comment somewhere to note that cs_vRecvMsg no longer protects vRecvMsg, and instead protects other stuff.

theuni commented:
From @sipa
> Can you clarify what the rationale is for switching from a ProcessMessages that processes multiple messages to just a single one? I'm not opposed to the change, but I'd like to understand what the reason for the change is.

In nearly all cases in the current code, only 1 message is processed. I thought that making this explicit would clarify the situation, but it seems it's only causing confusion in this PR. I'll change this back and we can address that later.

sipa commented:
@theuni Please ignore my request to keep processing multiple messages at the same time.

> In nearly all cases in the current code, only 1 message is processed.

I thought you were trying to say our send buffer was usually full, not that the code actually never loops (which is apparently behaviour that was intentionally changed in #3180).

theuni commented:
Updated to address feedback. Most of it is just fixing up interim commits to be more correct for easier commit-by-commit review.

Actual changes:
  - Fixed up the condvar as @sipa suggested
  - Completely removed cs_vRecvMsg. See the commit message for more detail.

Diff from before is here: https://gist.github.com/theuni/9c18405455a2569156671aadcf242500

morcos reviewed (commented):
Per IRC discussion, I think it would be wise to add something that shortens/skips the wait time at the end if the loop has taken a long time in particular for the case of connecting a new tip.  But that is a preexisting condition not a regression.

Overall looks good to me.
newbie quality code review ACK

theuni commented:
@morcos I meant to expand on the commit message for c5b4e31fb87a9819330b83f4ac6cc92066e8337d, but apparently neglected to do so before pushing. I expect that we'll want to add connman.WakeMessageHandler() in a few places to take care of that as a follow-up.

sipa commented:
utACK ddac068fa5d6b43925a96149f2d9b81ec736a5e7

pstratem commented on src/net.cpp:
```diff
@@ -648,6 +648,9 @@ void CNode::copyStats(CNodeStats &stats)
 bool CNode::ReceiveMsgBytes(const char *pch, unsigned int nBytes, bool& complete)
 {
     complete = false;
+    int64_t nTime = GetTimeMicros();
```
It would be nice if this was nTimeMicros instead of nTime

theuni replied:
will do

pstratem commented on src/net.cpp:
```diff
@@ -1882,10 +1878,10 @@ void CConnman::ThreadMessageHandler()
-            condMsgProc.wait_until(lock, std::chrono::steady_clock::now() + std::chrono::milliseconds(100));
-        }
+        std::unique_lock<std::mutex> lock(mutexMsgProc);
+        if (!fMoreWork)
```
Please add the {} back even though it's now 1 line

theuni replied:
will do

pstratem commented on src/net_processing.cpp:
```diff
@@ -2453,44 +2453,51 @@ bool ProcessMessages(CNode* pfrom, CConnman& connman, std::atomic<bool>& interru
 
     // this maintains the order of responses
-    if (!pfrom->vRecvGetData.empty()) return fOk;
+    if (!pfrom->vRecvGetData.empty()) return true;
```
Probably should swap these around.

theuni replied:
ok

theuni commented:
Addressed outstanding nits, will squash at the end.

laanwj commented:
(Lightly) tested ACK 1c779a8. This is an older version but it's been running stable on a busy node for a week.

theuni commented:
@TheBlueMatt @sipa @morcos care to re-ACK before final squash?

morcos commented:
utACK and ran on some nodes without error for several days ddac068
utACK 7b44e18

TheBlueMatt commented:
Dont think its actually possible due to the sizes involved, but, theoretically, PushMessage could increment nSendSize to be > nSendBufferMaxSize, set fPauseSend to true, and then call opportunistic SocketSendData, which flushed the entire buffer, but didnt reset fPauseSend. I'd suggest moving the fPauseSend update to right beside the nSendSize decrement as it is for the increment.

TheBlueMatt commented on src/net.cpp:
```diff
@@ -1239,20 +1239,16 @@ void CConnman::ThreadSocketHandler()
-                                WakeMessageHandler();
+                                completeMessages.splice(completeMessages.end(), pnode->vRecvMsg, pnode->vRecvMsg.begin(), it);
+                                if (!QueueReceivedMessages(pnode, std::move(completeMessages), nSizeAdded))
+                                    pnode->fPauseRecv = true;
```
By moving the setting of fPauseRecv outside of the cs_vProcessMsg lock you've introduced a race where the message processing thread may have already taken the next message and set fPauseRecv to false before this line (seems pretty unlikely, but its not unheard of). I'd suggest simply dropping this commit.

theuni replied:
Good catch!

TheBlueMatt commented:
utACK 7b44e18b2f3a39657699adff5dd39dcfde24f1fb if 975fc9dd6f55a87f7d0db62898e58169b2816baf is reverted and the fPauseSend updates are all placed directly beside the nSendSize updates.

TheBlueMatt commented:
utACK 0168bea73772a707c53952956a364262cea63cd3 (+/- squash-needed).

theuni commented:
Squashed, and I believe this is now merge-ready.

@sipa / @morcos See https://github.com/theuni/bitcoin/tree/connman-locks-tmp for the unsquashed version if you'd like. It's codewise-identical to e60360e139852c655930e99d4bb4db554cd8385e.

morcos commented:
re-utACK e60360e

sipa commented:
utACK e60360e139852c655930e99d4bb4db554cd8385e

rebroad commented on src/net_processing.cpp:
```diff
@@ -2463,48 +2450,48 @@ bool ProcessMessages(CNode* pfrom, CConnman& connman, std::atomic<bool>& interru
-            fOk = false;
-            break;
+            pfrom->fDisconnect = true;
+            return false;
```
why "false" and not "fMoreWork"?
